File: apps/parfast/mac/Sources/ParfastCore/BlockMapRule.swift
import Foundation

/// How a merged block-map cell picks its colour.
///
/// It lives here, as pure functions over a state array, rather than inside the
/// view, for two reasons: it is the one rendering rule the two apps must agree
/// on down to the pixel (it is THE signature visual), and it is the kind of
/// rule that is argued about from intuition unless somebody measures it.
///
/// THE MEASUREMENT, on the mock's ten-thousand-block set (188 damaged blocks,
/// 1.88%, scattered - which is what dropped articles look like) at the two
/// cell counts a real window produces:
///
/// | rule | 10,000 blocks, 188 scattered damaged | one lone bad block in 10,000 |
/// |---|---|---|
/// | worst state wins | **19.5% of the strip red** for 1.88% damage | 1 cell red |
/// | majority state only | 0% red - **damage invisible** | **0 cells red** |
/// | majority + bad tick | 0% red, 19.5% ticked | **1 cell ticked** |
///
/// So worst-wins over-states damage by an order of magnitude and majority
/// alone hides it completely; the third rule does neither, which is why it is
/// the one both apps implement. Ground for the PROPORTION, tick for the
/// PRESENCE, and the census under the map for the exact numbers.
///
/// Chip C (the Windows lane) argued for worst-wins on 12 Sep 2026, from the
/// correct premise that a merged cell must never hide a bad block. That
/// premise is what the tick satisfies. Do not "simplify" this back to one
/// rule without re-running the table above.
public enum BlockMapRule {

    /// How much a state matters when two are tied, and the order the hover
    /// readout lists them in. Not a severity ranking of the DATA - a ranking
    /// of what the user needs to see first.
    public static func rank(_ state: BlockState) -> Int {
        switch state {
        case .present: return 0
        case .pending: return 1
        case .hashing: return 2
        case .misnamed: return 3
        case .damaged: return 4
        case .missing: return 5
        }
    }

    /// A state is "bad" when it means data the set does not have where it
    /// expects it. MISNAMED IS NOT BAD: the data is on the disk under another
    /// name, it costs no recovery blocks, and painting it as damage makes a
    /// set two renames from perfect look nearly lost.
    public static func isBad(_ state: BlockState) -> Bool {
        switch state {
        case .damaged, .missing: return true
        case .present, .pending, .hashing, .misnamed: return false
        }
    }

    /// The half-open block range a merged cell covers.
    public static func range(column: Int, of columns: Int, blocks: Int) -> ClosedRange<Int> {
        let first = blocks * column / columns
        let last = max(first, blocks * (column + 1) / columns - 1)
        return first...min(last, blocks - 1)
    }

    /// The cell's ground colour: the most common state in its range, ties
    /// going to the one that matters more.
    public static func ground(_ states: [BlockState], in range: ClosedRange<Int>) -> BlockState {
        if range.lowerBound == range.upperBound { return states[range.lowerBound] }
        var counts = [Int](repeating: 0, count: BlockState.allCases.count)
        for i in range { counts[Int(states[i].rawValue)] += 1 }
        var best = BlockState.pending
        var bestCount = -1
        for state in BlockState.allCases {
            let n = counts[Int(state.rawValue)]
            if n > bestCount || (n == bestCount && rank(state) > rank(best)) {
                best = state
                bestCount = n
            }
        }
        return best
    }

    /// The tick along the bottom of the cell: the worst BAD state in the
    /// range, or nil when the range holds none. This is what stops the
    /// majority ground from hiding a single bad block in forty.
    public static func badMark(_ states: [BlockState], in range: ClosedRange<Int>) -> BlockState? {
        var worst: BlockState?
        for i in range {
            let s = states[i]
            guard isBad(s) else { continue }
            if worst == nil || rank(s) > rank(worst!) { worst = s }
        }
        return worst
    }

    /// How many blocks one merged cell covers, or nil when that is not yet
    /// knowable.
    ///
    /// NIL IS THE POINT. The caller's column count comes from a `@State`
    /// that `onAppear` fills in, so on the first render pass it is still
    /// zero - and the arithmetic that used to live at the call site,
    /// `blocks / max(1, columns)`, turns a zero column count into a
    /// per-cell figure of `blocks`. On the ten thousand block acceptance
    /// scenario that rendered as "each one covers 10,000 blocks" beside a
    /// map of about fourteen hundred cells: not a rounding error, the
    /// TOTAL, printed as if it were the per-cell count. Found 12 Sep 2026
    /// by the QA lane grading that scenario's screenshot against the
    /// corpus's measured numbers.
    ///
    /// A count is returned only when it can be computed, so a caller that
    /// has no columns yet shows NOTHING rather than something wrong. The
    /// floor of 2 is kept from the original: a merged map by definition
    /// has more blocks than cells, so a per-cell count of one means the
    /// division rounded down and two is the honest smallest answer.
    public static func blocksPerCell(blocks: Int, columns: Int) -> Int? {
        guard blocks > 0, columns > 0 else { return nil }
        return max(2, blocks / columns)
    }

    /// Minimum width, in POINTS, of any mark that says "something is wrong
    /// here". Chip C's finding, and it is the same defect as hiding the mark
    /// outright: a sub-pixel rectangle antialiases to nearly nothing, so a
    /// lone damaged cell can be drawn and still be invisible. Two points is
    /// two device pixels at 1x and four at 2x.
    public static let minimumMarkWidth: Double = 2
}

File: apps/parfast/windows/Parfast.ViewModels/BlockMapRule.cs
using Parfast.Core.Contracts;

namespace Parfast.ViewModels;

/// <summary>
/// How a merged block-map cell picks what it draws. A PORT, not a design.
/// </summary>
/// <remarks>
/// The rule is written down in <c>apps/parfast/shared/design/blockmap.md</c> and
/// the reference implementation is
/// <c>apps/parfast/mac/Sources/ParfastCore/BlockMapRule.swift</c>. This file is
/// that implementation in C#, function for function, so the two apps draw the
/// same picture of the same data. Do not re-derive it here; change the shared
/// document and both ports together.
/// <para>
/// THE MEASUREMENT THAT CHOSE IT, on the mock's ten-thousand-block set (188
/// damaged blocks, 1.88%, scattered, which is what dropped articles look like)
/// at 400 columns:
/// </para>
/// <list type="table">
/// <item><term>worst state wins</term><description>78 of 400 cells red, 19.5% of
/// the strip, for 1.88% damage</description></item>
/// <item><term>majority state only</term><description>0 cells red, the damage is
/// invisible</description></item>
/// <item><term>majority ground plus a bad tick</term><description>0 red, 78
/// ticked</description></item>
/// </list>
/// <para>
/// This lane argued for worst-wins on 12 Sep 2026, from the correct premise that
/// a merged cell must never hide a bad block. Chip B measured both and found the
/// premise right and the remedy wrong: worst-wins over-states damage by an order
/// of magnitude, so a repairable set looks a fifth destroyed on the screen the
/// user is deciding from. The TICK is what satisfies the premise without the
/// over-statement. Both failures are silent, which is why neither was going to be
/// caught by looking.
/// </para>
/// </remarks>
public static class BlockMapRule
{
    /// <summary>
    /// How much a state matters when two are tied, and the order the hover
    /// readout lists them in. Not a severity ranking of the DATA, a ranking of
    /// what the user needs to see first.
    /// </summary>
    public static int Rank(BlockState state) => state switch
    {
        BlockState.Present => 0,
        BlockState.Pending => 1,
        BlockState.Hashing => 2,
        BlockState.Misnamed => 3,
        BlockState.Damaged => 4,
        BlockState.Missing => 5,
        _ => 0,
    };

    /// <summary>
    /// A state is bad when it means data the set does not have where it expects
    /// it. MISNAMED IS NOT BAD: the data is on the disk under another name, it
    /// costs no recovery blocks, and painting it as damage makes a set two
    /// renames from perfect look nearly lost.
    /// </summary>
    public static bool IsBad(BlockState state) =>
        state is BlockState.Damaged or BlockState.Missing;

    /// <summary>The half-open block range a merged cell covers: [First, Last].</summary>
    public static (int First, int Last) Range(int column, int columns, int blocks)
    {
        var first = (int)((long)blocks * column / columns);
        var last = Math.Max(first, (int)((long)blocks * (column + 1) / columns) - 1);
        return (first, Math.Min(last, blocks - 1));
    }

    /// <summary>
    /// The cell's ground: the most common state in its range, ties going to the
    /// one that matters more.
    /// </summary>
    public static BlockState Ground(IReadOnlyList<BlockState> states, int first, int last)
    {
        if (first == last)
        {
            return states[first];
        }

        Span<int> counts = stackalloc int[8];
        for (var i = first; i <= last; i++)
        {
            var code = (int)states[i];
            if ((uint)code < (uint)counts.Length)
            {
                counts[code]++;
            }
        }

        var best = BlockState.Pending;
        var bestCount = -1;
        foreach (var state in All)
        {
            var n = counts[(int)state];
            if (n > bestCount || (n == bestCount && Rank(state) > Rank(best)))
            {
                best = state;
                bestCount = n;
            }
        }

        return best;
    }

    /// <summary>
    /// The tick along the bottom of the cell: the worst BAD state in the range,
    /// or null when it holds none. This is what stops the majority ground from
    /// hiding a single bad block in forty.
    /// </summary>
    public static BlockState? BadMark(IReadOnlyList<BlockState> states, int first, int last)
    {
        BlockState? worst = null;
        for (var i = first; i <= last; i++)
        {
            var s = states[i];
            if (!IsBad(s))
            {
                continue;
            }

            if (worst is null || Rank(s) > Rank(worst.Value))
            {
                worst = s;
            }
        }

        return worst;
    }

    /// <summary>
    /// Minimum width, in device-independent pixels, of any mark that says
    /// something is wrong here. This lane's finding, and it is the same defect as
    /// not drawing the mark at all: a sub-pixel rectangle antialiases to nearly
    /// nothing, so a lone damaged cell can be drawn and still be invisible.
    /// </summary>
    public const double MinimumMarkWidth = 2;

    internal static readonly BlockState[] All =
    [
        BlockState.Pending, BlockState.Present, BlockState.Damaged,
        BlockState.Missing, BlockState.Misnamed, BlockState.Hashing,
    ];
}
